torchao/prototype/low_bit_optim/cpu_offload.py

```diff
@@ -0,0 +1,34 @@
+import torch
+from torch.optim.optimizer import Optimizer
+
+
+class CPUOffloadOptimizer:
+    def __init__(self, base_optimizer: Optimizer) -> None:
+        self.optim = base_optimizer
+        self.param_cpu2cuda_map = dict()
+
+        # swap param in param_groups with CPU param
+        for param_group in base_optimizer.param_groups:
+            for i, p in enumerate(param_group["params"]):
+                p_cpu = p.detach().cpu().pin_memory()
+                param_group["params"][i] = p_cpu
+                self.param_cpu2cuda_map[p_cpu] = p
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        # copy gradients from CUDA to CPU
+        for p_cpu, p_cuda in self.param_cpu2cuda_map.items():
+            if p_cuda.grad is not None:
+                p_cpu.grad = p_cuda.grad.to("cpu", non_blocking=True)
```

Comment:
To check that we on the same page, the `non_blocking=True` here means that the host (CPU) is not blocked on this D2H copy. However, there is nothing for these D2H copies to overlap with, so the main benefit you are getting here is that copying D2H with `non_blocking=True` will copy directly to pinned memory.

Otherwise, the CPU side should look like issuing D2H copy for each gradient and then blocking via the `torch.cuda.synchronize()` for all D2H copies to finish.